Ballz improvements

BlackMinou opened the pull request:
Better concern division into modules, now works on Linux ! 
forces.nit module is not really used yet but will be.

PS : mnit_simple and mnit_ballz aren't launching on android at the moment, I'll fix this in a further PR

Edit: Fixed Android part

Morriar commented:
Not sure how this issue's name passed through my spam filters.

Morriar commented on examples/mnit_ballz/src/assets.nit:
```diff
@@ -0,0 +1,49 @@
+import mnit::assets
+
+
+# Conttains all the assets
```
Contains*

Morriar commented on examples/mnit_ballz/src/assets.nit:
```diff
@@ -0,0 +1,49 @@
+
+
+# Conttains all the assets
+class Assets
```
I think your api would be prettier using prefixes to dissociate assets:

* snd_wall_destr
* snd_bounce
* img_ball
* img_hwall
* img_vwall

But it's only my OCD talking at this point.

Morriar commented on examples/mnit_ballz/src/assets.nit:
```diff
@@ -0,0 +1,49 @@
+	var vertical_wall: Image is noinit
+
+	# Loads all the assets
+	fun load do
```
Just for my information: is there any reasons why the APIs of sound and image loading are that different?

privat replied:
Agree. Image loading should use something like the sound API that seems cleaner.

Morriar commented on examples/mnit_ballz/src/ballz_linux.nit:
```diff
@@ -0,0 +1,126 @@
+	do
+		var value = 5.0
+		if event isa SDLKeyEvent then
+			if event.is_down then
```
Can't it be factorized like

~~~nit
if event.is_arrow_left then
+					acceleration(value, 0.0)
+					app.left_arrow_down = event.is_down
+				end
+				if event.is_arrow_right then
+					acceleration(-value, 0.0)
+					app.right_arrow_down = event.is_down
+				end
+				if event.is_arrow_up then
+					acceleration(0.0, -value)
+					app.up_arrow_down = event.is_down
+				end
+				if event.is_arrow_down then
+					acceleration(0.0, value)
+					app.down_arrow_down = event.is_down
+				end
~~~

Morriar commented on examples/mnit_ballz/src/collision.nit:
```diff
@@ -0,0 +1,90 @@
+	return ( x * x + y * y ).sqrt
+end
+
+# Get the magnitude of `vector` ( = its length )
```
~~~nit
# Get the magnitude (length) of `vector`
~~~

Morriar commented on examples/mnit_ballz/src/collision.nit:
```diff
@@ -0,0 +1,90 @@
+# Get the magnitude of `vector` ( = its length )
+fun magnitude(vector: Point[Float]): Float do return ( vector.x * vector.x + vector.y * vector.y ).sqrt
+
+# Get the unit vector of `vector`
```
One word about what is a `unit` vector?

privat replied:
https://en.wikipedia.org/wiki/Unit_vector :)

Morriar commented on examples/mnit_ballz/src/collision.nit:
```diff
@@ -0,0 +1,90 @@
+# Get the dot product of vectors `v1` and `v2`
+fun dot_product(v1, v2: Point[Float]): Float do return v1.x * v2.x + v1.y * v2.y
+
+# Get the vector between start point `sp` and end point `ep`
```
With a better signature:

~~~nit
# Get the vector between `start_point` and `end_point`
~~~

Morriar commented on examples/mnit_ballz/src/collision.nit:
```diff
@@ -0,0 +1,90 @@
+	return distance < radius
+end
+
+# bouncing function, return the new point of the center of the ball
```
Please capitalize that `bouncing`.

Also, written this way I think it's `returns`

Morriar commented on examples/mnit_ballz/src/display.nit:
```diff
@@ -0,0 +1,47 @@
+
+redef class Game
+
+	# Draw all the entities onto `display`
```
Entities or `assets`?

BlackMinou replied:
I'd say entities since each entities knows its own assets, the `draw` function in `Game` just delegate the drawing to the entities, even if ultimately they draw their assets on the screen

Morriar commented on examples/mnit_ballz/src/forces.nit:
```diff
@@ -0,0 +1,34 @@
+# This module contains all the forces applicables to the objects in the game
```
all the applicable forces*

privat replied:
The "This module contains" part is useless. Maybe "Physical model for the applicable forces to the objects in the game"

Morriar commented on examples/mnit_ballz/src/game_logic.nit:
```diff
@@ -14,167 +14,190 @@
-			print "ASensorProximity : distance = " + event.distance.to_s
-		else if event isa MotionEvent then
-			activate_walls(event)
+	# Apply all the forces applicables to `self`
```
see `applicable forces`

Morriar commented on examples/mnit_ballz/src/game_logic.nit:
```diff
@@ -14,167 +14,190 @@
-		ball_img.scale = scale
-		ball_img.scale = 3.0
-	end
+	# The Ball !
```
`The Ball!`

Morriar commented on examples/mnit_ballz/src/game_logic.nit:
```diff
@@ -14,167 +14,190 @@
-		display.blit_rotated(ball_img, game.ball.x, game.ball.y, 0.0)
-	end
+	# List of walls in the level
+	var walls: Array[Wall] is noinit
```
Is the order of insertion relevant?

BlackMinou replied:
Not really, why ?

Morriar commented on examples/mnit_ballz/src/objects.nit:
```diff
@@ -0,0 +1,90 @@
+import geometry::boxes
+import collision
+
+# The ball is the main character of the game !
```
No space before the exclamation mark.

Morriar commented on examples/mnit_ballz/src/objects.nit:
```diff
@@ -0,0 +1,90 @@
+
+# The ball is the main character of the game !
+class Ball
+	#center of the ball
```
~~~nit
# Center of the ball
~~~

privat commented:
jenkins: ok to test

privat commented on examples/mnit_ballz/Makefile:
```diff
@@ -7,6 +7,10 @@ android: icon
 	mkdir -p bin
 	../../bin/nitc -o bin/ballz.apk src/ballz_android.nit
 
+linux:
```
maybe build linux by default now

BlackMinou replied:
Done

privat commented on examples/mnit_ballz/src/assets.nit:
```diff
@@ -0,0 +1,49 @@
+class Assets
+
+	# Sound for the wall destruction
+	var wall_destruction = new Sound("walld")
```
The sound does not remind me anything related to wall destruction.

BlackMinou replied:
Are you talking about the name of the sound ? Or the fact that there is no longer wall destruction implemented in the game ?

privat replied:
I just played the sound, it sounded weird.

BlackMinou replied:
Yeah I took theses sounds a long ago from the linux system xD

privat commented on examples/mnit_ballz/src/assets.nit:
```diff
@@ -0,0 +1,49 @@
+	var wall_destruction = new Sound("walld")
+
+	# Sound when the ball bounce
+	var bounce = new Sound("bounce")
```
Coin obtained? You should check your sounds. Moreover the 2sec of silence at the end of the file.

privat replied:
By the way, why one is a ogg/vorbis while the other is a pcm/wav?

BlackMinou replied:
testing purpose, since mnit_ballz is in example, you can see that different formats can be use

xymus replied:
Does it really work without the extensions? I guess it does on Android from the res folder, but on Linux too?

BlackMinou replied:
There is a little problem here : 
It doesn't work because on linux because of the extension, that's for sure.

The problem is, the android Sound API tries to load a sound from assets first, and it needs the file extension. It then tries to load a sound from the res/raw folder, and it doesn't work with the extension on the file. I thought about using `strip_extension` when it comes to load the sound from the resources, but then I walked into another problem : 
Linux assets are located in the `assets` folder, nothing abnormal here. Android assets are in the assets folder or the resources arborescence. So when you want to make an application that runson both android and Linux, but you want to use the resources on android, you'll always be loading the sounds from the assets folder since the Android Sound API tries it first.

Got 2 solutions for that : 
* Reverse the loading order of the sound in the Android API
* Just use the `strip_extension` trick, but then you'll never load your sounds from somewhere else than the assets folder when you're making an application that runs on both linux and android

Well after writing all of that, I think i'll just use the first solution \o/

privat commented on examples/mnit_ballz/src/ballz_android.nit:
```diff
@@ -14,21 +14,30 @@
+	app_version(0, 2, git_revision)
 	app_name("Ballz")
+	app_namespace "org.nitlanguage.ballz"
+	android_api_target 20
```
API 20 is 4.4W.2 Something wear-related. Are you sure?

BlackMinou replied:
My local environment can only compile for API 20 or +, that's why it's here

privat replied:
Improve your local environment then.

privat commented on examples/mnit_ballz/src/ballz_android.nit:
```diff
@@ -14,21 +14,30 @@
 	app_name("Ballz")
+	app_namespace "org.nitlanguage.ballz"
+	android_api_target 20
+	android_manifest("""<uses-permission android:name="android.permission.INTERACT_ACROSS_USERS" />""")
```
Do you plan to violate the protection between users?

BlackMinou replied:
Was more a debug line, will remove

privat commented on examples/mnit_ballz/src/ballz_android.nit:
```diff
@@ -14,21 +14,30 @@
-import game_logic
+import android::sensors
+import android::audio
+import android::assets_and_resources
```
Thanks to #1580, android::audio and android::assets_and_resources should be automatically loaded

privat commented on examples/mnit_ballz/src/ballz_linux.nit:
```diff
@@ -0,0 +1,126 @@
+module ballz_linux
+
+import mnit_linux
+import linux::audio
```
Thanks to #1580, linux::audio should be automatically loaded

privat commented on examples/mnit_ballz/src/ballz_linux.nit:
```diff
@@ -0,0 +1,126 @@
+				end
+			end
+		end
+	return false
```
bad indentation

privat commented on examples/mnit_ballz/src/display.nit:
```diff
@@ -0,0 +1,47 @@
+	# Draw `self` onto `display` with image from `assets`
+	fun draw(display: Display, assets: Assets)
+	do
+		display.blit_rotated(assets.ball, center.x, center.y, 0.0)
```
what is the point of `blit_rotated` with an angle of 0.0?

BlackMinou replied:
Not sure but if I replace it with `blit`, it fucks everything up ...

xymus replied:
It would be `blit_centered`, the coordinates in `blit` are for the top left corner. But there's no real performance difference on OpenGL between these 2 methods.

privat commented on examples/mnit_ballz/src/display.nit:
```diff
@@ -0,0 +1,47 @@
+redef class Wall
+
+	# Draw `self` onto `display` with image from `assets`
+	fun draw(display: Display, assets: Assets)
```
Maybe you should have a super-class of Wall and Ball to abstract services like `draw`. And possibly also `Game` if you plan a Composite Design Pattern

BlackMinou replied:
Not yet, not enough abstraction needed

privat commented on examples/mnit_ballz/src/game_logic.nit:
```diff
@@ -14,167 +14,190 @@
-import android::sensors
+import assets
+import objects
+import forces
```
TODO: rename `import` as `use` to enable a pun.

privat commented on examples/mnit_ballz/src/forces.nit:
```diff
@@ -0,0 +1,34 @@
+end
+
+# Gravity applying to objects
+class Gravity
```
unused class

BlackMinou replied:
the module is unused for now so yeah this class too

privat commented on examples/mnit_ballz/src/forces.nit:
```diff
@@ -0,0 +1,34 @@
+
+	redef var application = new Point[Float](0.85, 0.85)
+
+	redef fun apply(p) do return new Point[Float](p.x * application.x, p.y * application.y)
```
Warning: you redefined a method with the exact same body

BlackMinou replied:
nope, if you look closer, the `+` operator became a `*`

privat replied:
Quite tricky, please document it.

privat commented on examples/mnit_ballz/src/game_logic.nit:
```diff
@@ -14,167 +14,190 @@
+		var np = collision(game.quadtree)
+		if np != null then
+				offset = np
+				center = new Point[Float](center.x + offset.x, center.y + offset.y)
```
bad indentation

privat commented on examples/mnit_ballz/src/game_logic.nit:
```diff
@@ -14,167 +14,190 @@
+		var wallb = new Wall.complete(new Point[Float](width*0.75, height/4.0), 0.0, 1.0)
+		var wallc = new Wall.complete(new Point[Float](width/4.0, height*0.75), 0.0, 1.0)
+		var walld = new Wall.complete(new Point[Float](width*0.75, height*0.75), pi/3.0, 1.0)
+		walls = new Array[Wall].with_items(walla, wallb, wallc, walld)
```
I like the level design

BlackMinou replied:
level serialization for a future version ...

privat commented on examples/mnit_ballz/src/objects.nit:
```diff
@@ -0,0 +1,90 @@
+end
+
+# Walls make the ball bounce on them
+class Wall
```
Are walls non-playable characters?

BlackMinou replied:
yes, they're not playable, but it could be fun xD, will note that somewhere thx for the idea

privat replied:
Can walls have feelings, too?

privat commented on examples/mnit_ballz/src/objects.nit:
```diff
@@ -0,0 +1,90 @@
+	var center: Point[Float]
+
+	# angle in degrees
+	var angle: Float
```
degree or radiant?

privat commented on examples/mnit_ballz/src/objects.nit:
```diff
@@ -0,0 +1,90 @@
+	var lines: Array[OrientedLine[Float]] is noautoinit
+
+	# Initialize `self` with all its lines from `center` and `angle`
+	init complete(center: Point[Float], angle: Float,  scale: Float) do
```
why a named constructor here? Remove `noautoinit` in center, angle and scale and win a free autoinit constructor

privat commented on examples/mnit_ballz/src/game_logic.nit:
```diff
@@ -14,167 +14,190 @@
+	# Quadtree used for collision detection
+	var quadtree: SQuadTree[OrientedLine[Float]]
 
 	init(display: Display)
```
Old init style is old.
And bad OO design as you add unneeded coupling Game -> Display here.
Use a single autoinit constructor by removing `noautoinit` from width and height.

privat commented on examples/mnit_ballz/src/objects.nit:
```diff
@@ -0,0 +1,90 @@
+# A line represented with a center and an angle
+class OrientedLine[N]
+	super Line[Numeric]
+	redef type P: Point[Float]
```
`N` is unused, and the conjoint use of Numeric and Float make me feel uneasy. please stop doing that.

BlackMinou replied:
sorry, did this one really fast without thinking enough -_-

privat commented:
Since android is borken, maybe remove anything related to android in the PR. You will introduce them in a next one.

R4PaSs commented on examples/mnit_ballz/src/ballz_android.nit:
```diff
@@ -38,32 +44,65 @@ redef class App
+	do
+		if event isa ASensorAccelerometer then
+			acceleration(event.x, event.y)
+		else if event isa ASensorMagneticField then
```
Why all the commented parts ? I know this is an example and it's supposed to show a little of how it's like to use sensors, but is it necessary to leave these many commentaries ?

BlackMinou replied:
Yes, and this code may also be useful for next pull requests if other events are used

R4PaSs commented on examples/mnit_ballz/src/collision.nit:
```diff
@@ -0,0 +1,90 @@
+# Get the vector between `start_point` and `end_point`
+fun vector_between(start_point, end_point: Point[Float]): Point[Float] do return new Point[Float](end_point.x - start_point.x, end_point.y - start_point.y)
+
+
```
Why the two `\n` ?

R4PaSs commented on examples/mnit_ballz/src/game_logic.nit:
```diff
@@ -14,167 +14,188 @@
-		else if event isa MotionEvent then
-			activate_walls(event)
+	# Apply all the applicable forces to `self`
+	# Not uset yet
```
:s/uset/used

xymus replied:
@BlackMinou , you should `TODO` in such cases.

xymus commented on examples/mnit_ballz/src/collision.nit:
```diff
@@ -0,0 +1,90 @@
+
+
+# Returns the point on a line with endpoints `l1` and `l2` closest to the `center` of the ball
+fun point_on_line_closest_to_ball(center, l1, l2: Point[Float]): Point[Float]
```
We can't argue against methods with such a precise name!

xymus commented on examples/mnit_ballz/src/game_logic.nit:
```diff
@@ -14,167 +14,188 @@
 end
+
+# Get the assets of the app
+fun assets: Assets do return app.assets
```
This looks unused,  you use `assets` only in `App` or in a `draw(assets...)`. It can probably be removed.

xymus commented:
Finally we have the much hyped walls, and as a bonus it's a nice clean up!

Unless you plan on completing the features related to the dead code/module right away, you should probably remove it. Keep it for a future commit and it will kick start you next PR!

Would you mind standardizing you commit messages? It will be easier to follow once integrated if they all begin with the same prefix!

privat commented on examples/mnit_ballz/src/game_logic.nit:
```diff
@@ -14,167 +14,172 @@
+redef class App
+
+	# Assets used in all the app
+	var assets = new Assets
```
logic issue here: `new Assets` will require things from `app` but because you are in the default value of an attribute of App, the object is not initialized yet thus `app` is not assigned. Therefore that will do an infinite recursion that will segfault the program.

BlackMinou replied:
fixed

BlackMinou commented:
Reroll : 
* Standardized commits
* fixed commentaries
* removed the forces part (sorry for your pun @privat)
* fixed android part, was an infinite recursion in App creation :/

xymus commented:
:shipit: !
